wgpu-core/src/hub.rs

```diff
@@ -446,93 +479,151 @@ impl<B: GfxBackend, F: GlobalIdentityHandlerFactory> Hub<B, F> {
+                            }
                         }
+                        TextureViewInner::SwapChain { .. } => {} //TODO
                     }
-                    TextureViewInner::SwapChain { .. } => {} //TODO
                 }
             }
         }
 
-        for (_, (texture, _)) in self.textures.data.write().map.drain() {
-            devices[texture.device_id.value].destroy_texture(texture);
-        }
-        for (_, (buffer, _)) in self.buffers.data.write().map.drain() {
-            //TODO: unmap if needed
-            devices[buffer.device_id.value].destroy_buffer(buffer);
-        }
-        for (_, (command_buffer, _)) in self.command_buffers.data.write().map.drain() {
-            devices[command_buffer.device_id.value]
-                .com_allocator
-                .after_submit(command_buffer, 0);
-        }
-        for (_, (bind_group, _)) in self.bind_groups.data.write().map.drain() {
-            let device = &devices[bind_group.device_id.value];
-            device.destroy_bind_group(bind_group);
-        }
-
-        for (_, (module, _)) in self.shader_modules.data.write().map.drain() {
-            let device = &devices[module.device_id.value];
-            unsafe {
-                device.raw.destroy_shader_module(module.raw);
-            }
-        }
-        for (_, (bgl, _)) in self.bind_group_layouts.data.write().map.drain() {
-            let device = &devices[bgl.device_id.value];
-            unsafe {
-                device.raw.destroy_descriptor_set_layout(bgl.raw);
-            }
-        }
-        for (_, (pipeline_layout, _)) in self.pipeline_layouts.data.write().map.drain() {
-            let device = &devices[pipeline_layout.device_id.value];
-            unsafe {
-                device.raw.destroy_pipeline_layout(pipeline_layout.raw);
+        self.textures.data.write().map.drain(..).for_each(|x| {
+            if let Element::Occupied(texture, _) = x {
+                devices[texture.device_id.value].destroy_texture(texture);
             }
-        }
-        for (_, (pipeline, _)) in self.compute_pipelines.data.write().map.drain() {
-            let device = &devices[pipeline.device_id.value];
-            unsafe {
-                device.raw.destroy_compute_pipeline(pipeline.raw);
+        });
+        self.buffers.data.write().map.drain(..).for_each(|x| {
+            if let Element::Occupied(buffer, _) = x {
+                //TODO: unmap if needed
+                devices[buffer.device_id.value].destroy_buffer(buffer);
             }
-        }
-        for (_, (pipeline, _)) in self.render_pipelines.data.write().map.drain() {
-            let device = &devices[pipeline.device_id.value];
-            unsafe {
-                device.raw.destroy_graphics_pipeline(pipeline.raw);
+        });
+        self.command_buffers
+            .data
+            .write()
+            .map
+            .drain(..)
+            .for_each(|x| {
```

Comment:
is there any reason to use `for_each` in all these constructs? I think the code is easier if it's just a for loop. Less closures involved.